### ISSUE TITLE
Environments should reference a cue module, instead of embedding one

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#631]
## Proposal

Instead of each Dagger environment embedding a complete copy of a cue module (`.dagger/env/FOO/plan/*.cue`), add a new field `cue` to the environment configuration (`.dagger/env/FOO/values.yaml`):

* `plan: module` (optional) references the local path of the cue module to use as a plan, relative to the workspace root. Default: `.`.
* `plan: package` (optional) references the name of the Cue package within the module to use as a plan. Default: defer to cue loader (accept any unique package name, or error if multiple packages exist).

NOTE: it is important that `plan: module` always be a local path. One must resist the temptation to support, eg. remote references to cue modules. This is because local module directories *contain dependency management information*. By referencing a local cue module (and requiring that a cue module exist), we effectively offload dependency management to Cue. If we were to reference a remote Cue module, Dagger itself would need to answer the question “at which version?”. This would result in creating a parallel dependency management system

## Rationale

The experience of developing for Dagger should be as Cue-native as possible. For example, when reading third-party Cue development resources (unrelated to Dagger), a Dagger developer should be comfortable and know how to use those resources in their Dagger development setup. Conversely, an experienced Cue developer should recognize the Dagger development workflow and be comfortable adopting it without losing productivity.

To achieve this, it is necessary for [Cue modules]() to be visible and at their expected location. Currently that is not the case: each Dagger environment *embeds* a cue module as their “delivery plan”. For example an environment called “staging” will embed a cue module at `.dagger/env/staging/plan`. The module’s `cue.mod` will be found at `.dagger/env/staging/plan/cue.mod`. This poses no problem when loading Cue (whether `cue eval` or `dagger up`). But it complicates *dependency management* which is an important part of the development process. This is because a Cue module, like a Go module, has a name (stored in `cue.mod/module.cue`). And that name is expected to be externally valid, so that third-party configurations may import the module as a dependency. In the common case of the package name being tied to the git hosting infrastructure (for example `github.com/foo/bar` or `gitlab.com/hello/world`), this is problematic because developers must now choose between A) a human-friendly name which breaks dependency management (`github.com/foo/bar` is not the actual location), or B) an ugly name which respects Cue dependency management but is long to type, and eposes implementation details of Dagger’s file layout (`github.com/foo/bar/.dagger/env/bar/plan`).

Moving the Cue module out of the `.dagger` directory solves this.

## Note on other possible changes

There are several other possible changes to the environment layout I would like to propose. For example:

* `values.yaml` is a confusing filename inherited from initial proof-of-concept implementation, which we never found the time to change
* The current environment naming system (`name` field in `values.yaml`) is inherited from an earlier layout where environments were not organized in a flat namespace of sub-directories. Now that they are, the field is redundant.
* As discussed with @proppy and @myitcv , it would be possible to go even further in the direction of “cue native”, and for example move the contents of `values.yaml` into a non-hidden file, embedded in the cue module. Copying the simplicity of Waypoint, we could for example combine all workspace configuration into a single `dagger.yaml`. At which point it would be embarrassing not to call it `dagger.cue`. At that point, `.dagger` could be reserved simply for state and anything not to be committed in git.

All of these proposals are interesting (and even exciting) but they are out of scope for this issue. The goal is to focus on the most urgently needed change first, while continuing to discuss other possible changes in parallel. Copying @proppy, @myitcv and @dubo-dubon-duponey for those :)
